File: backend/app/schemas/session.py
"""会话 API 的 Pydantic 模型。参照 BACKEND_STRUCTURE §2.4。"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator


class SessionCreate(BaseModel):
    story_id: int
    mode: str = Field(..., pattern="^(strict|creative)$")
    opening_goal: str = Field(..., min_length=1, max_length=8000)
    rag_config_id: int | None = None
    style_config: dict[str, Any] | None = None


class SessionMessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    turn_number: int
    role: str
    content: str
    metadata: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="before")
    @classmethod
    def _coerce_orm(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return data
        if hasattr(data, "metadata_"):
            return {
                "id": data.id,
                "turn_number": data.turn_number,
                "role": data.role,
                "content": data.content,
                "metadata": data.metadata_ if data.metadata_ is not None else {},
            }
        return data


class SessionStateSnapshot(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    turn_number: int
    state: dict[str, Any]
    created_at: datetime | None = None


class SessionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    story_id: int
    story_version_id: int
    rag_config_id: int
    mode: str
    opening_goal: str
    style_config: dict[str, Any]
    status: str
    turn_count: int
    created_at: datetime | None = None
    updated_at: datetime | None = None
    latest_state: SessionStateSnapshot | None = None


class SessionListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    story_id: int
    story_version_id: int
    rag_config_id: int
    mode: str
    status: str
    turn_count: int
    opening_goal: str
    created_at: datetime | None = None
    updated_at: datetime | None = None


class FeedbackCreate(BaseModel):
    message_id: int
    feedback_type: str = Field(..., min_length=1, max_length=30)
    content: str | None = Field(None, max_length=4000)


class UserFeedbackOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: int
    message_id: int
    feedback_type: str
    content: str | None
    reviewed: bool
    created_at: datetime | None = None


class SessionCreateResponse(SessionResponse):
    opening_message: SessionMessageOut | None = None


class SessionMessageCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=12000)


class OpeningGenerationResponse(BaseModel):
    narrative: str
    choices: list[str]
    state_update: dict[str, Any]
    parse_error: str | None = None

File: frontend/src/types/session.ts
/** 与后端 session schemas 对齐 */

export type SessionMode = "strict" | "creative";

export interface SessionMessage {
  id: number;
  turn_number: number;
  role: "user" | "assistant";
  content: string;
  metadata: Record<string, unknown>;
}

export interface SessionStateSnapshot {
  id: number;
  turn_number: number;
  state: NarrativeState;
  created_at?: string | null;
}

/** 叙事状态 JSON（与后端 state.initialize_state 一致） */
export interface NarrativeState {
  current_location?: string;
  active_goal?: string;
  important_items?: unknown[];
  npc_relations?: Record<string, string>;
}

/** 与后端 SessionListItem 对齐（GET /api/sessions） */
export interface SessionListItem {
  id: number;
  story_id: number;
  story_version_id: number;
  rag_config_id: number;
  mode: string;
  status: string;
  turn_count: number;
  opening_goal: string;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface SessionResponse {
  id: number;
  user_id: number;
  story_id: number;
  story_version_id: number;
  rag_config_id: number;
  mode: string;
  opening_goal: string;
  style_config: Record<string, unknown>;
  status: string;
  turn_count: number;
  created_at?: string | null;
  updated_at?: string | null;
  latest_state: SessionStateSnapshot | null;
}

export interface SessionCreatePayload {
  story_id: number;
  mode: SessionMode;
  opening_goal: string;
  rag_config_id?: number | null;
  style_config?: Record<string, unknown> | null;
}

export interface OpeningGenerationResponse {
  narrative: string;
  choices: string[];
  state_update: Record<string, unknown>;
  parse_error: string | null;
}

export interface FeedbackPayload {
  message_id: number;
  feedback_type: string;
  content?: string | null;
}
